cpp/dolfinx/fem/interpolate.h

```diff
@@ -640,4 +645,43 @@ void interpolate(Function<T>& u, const Function<T>& v)
   }
 }
 
+/// Interpolate from a dolfinx Expression into a compatible Function on the same
+/// mesh
+/// @param[in, out] u The function to interpolate into
+/// @param[in] expr The expression to be interpolated
+/// @param[in] cells List of cell indices (local to process) to interpolate on
+template <typename T>
+void interpolate(Function<T>& u, const Expression<T>& expr,
+                 const tcb::span<const std::int32_t>& cells)
+{
+  // Array to hold coefficients after expression evaluation
+  const std::int32_t num_points = expr.num_points();
+  const std::int32_t expr_value_size = expr.value_size();
+  const std::int32_t num_cells = cells.size();
+  xt::xtensor<T, 2> expr_values(
+      {std::size_t(cells.size()), std::size_t(num_points * expr_value_size)});
+
+  // Check that spaces are compatible
+  std::shared_ptr<const FiniteElement> element = u.function_space()->element();
+  const auto dofmap = u.function_space()->dofmap();
+  assert(dofmap);
+  const int element_vs = element->value_size();
+  assert(expr_value_size == element_vs);
+  auto X = expr.x();
+  assert(X.shape() == element->interpolation_points().shape());
+
+  // Evaluate expression at quadrature points for the subset of cells
+  expr.eval(cells, expr_values);
+
+  // Reshape evaluated data to fit interpolate
```

Comment:
Can this be simplified, or the doc string expanded? It’s hard to follow the multiple reshaping steps.

Reply:
Expression takes in a matrix of shape (num_cells, num_points*value_size), where the last dimension varies with xyzxyz.
Interpolate takes in a matrix of shape (value_size, num_cells*num_points), i.e xxyyzz ordering where all points of cell0 comes first, then cell1 etc.

Reply:
This is still confusing - why is the reshaping done in more than one step?

Reply:
How would you do it in one step?